### PR TITLE
feat(admin): sponsor this member

### PR DESCRIPTION
**Pull request by sarkapalkovicova**
Enable sponsor members from their profile.

### CODE REVIEW

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts [hunk 1]
```diff
@@ -224,4 +225,32 @@ export class MemberOverviewComponent implements OnInit {
...
+
+    const dialogRef = this.dialog.open(SponsorThisMemberDialogComponent, config);
+
+    dialogRef.afterClosed().subscribe((edited) => {
+      if (edited) {
+        this.loading = true;
+        this.membersService.getRichMemberWithAttributes(this.member.id).subscribe((member) => {
+          this.member = member;
+          if (this.member.sponsored) {
+            this.usersManager.getSponsorsForMember(this.member.id, null).subscribe((sponsors) => {
```
Comment (bodnara) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts:246]: Here we are sure the member is sponsored (result of the dialog actions) and the check can be omitted.

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts [hunk 2]
```diff
@@ -224,4 +225,32 @@ export class MemberOverviewComponent implements OnInit {
       () => (this.loading = false)
     );
   }
+
+  private sponsorThisMember(): void {
```
Comment (bodnara) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts:229]:
```suggestion
  sponsorThisMember(): void {
```

File: apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.components.html [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    <div>
+      <br />
+      <h6>{{'DIALOGS.SPONSOR_THIS_MEMBER.EXPIRATION' | translate}}</h6>
+      <perun-web-apps-expiration-select [minDate]="minDate" (datePicker)="setExpiration($event)">
+      </perun-web-apps-expiration-select>
+    </div>
+  </div>
+
+  <div mat-dialog-actions>
+    <button (click)="onCancel()" class="ms-auto" mat-flat-button>
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.components.html:20]:
```suggestion
    <button (click)="onCancel()" class="ms-auto" mat-stroked-button>
```

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html [hunk 4]
```diff
@@ -68,17 +68,28 @@ <h1 class="page-subtitle">{{'MEMBER_DETAIL.OVERVIEW.RESET_PASSWORD' | translate}
...
+
+          <mat-card-content *ngIf="!member.sponsored" class="column-center">
+            <button
+              (click)="sponsorThisMember()"
+              class="action-button"
+              color="accent"
+              mat-flat-button>
+              {{'MEMBER_DETAIL.OVERVIEW.SPONSOR_MEMBER'|translate}}
+            </button>
+          </mat-card-content>
```
Comment (bodnara) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html:90]: The button should be visible also if member is already sponsored.

**Review by bodnara — CHANGES_REQUESTED** [review]

**Review by bodnara — CHANGES_REQUESTED** [review]
We should try to prevent `AlreadySponsorException`, in some sensible manner. So for just sponsors hide the button if they already sponsor this member. Otherwise filter the remaining sponsors in organization. There are edge cases here (self is sponsoring but can add other and all sponsors sponsor the member) that I am not sure how to handle the best right now.
Also the search field expect `User` to be passed not `Sponsor` which I find a bit misleading but now the search shows just:
image

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html [hunk 5]
```diff
@@ -68,17 +68,29 @@ <h1 class="page-subtitle">{{'MEMBER_DETAIL.OVERVIEW.RESET_PASSWORD' | translate}
...
+          <mat-card-title>
+            <h1 class="page-subtitle">{{'MEMBER_DETAIL.OVERVIEW.SPONSORSHIP' | translate}}</h1>
+          </mat-card-title>
+          <mat-card-subtitle class="mt-n3 mb-2" *ngIf="member.sponsored">
+            {{'MEMBER_DETAIL.OVERVIEW.SPONSORED_BY' | translate}}
+          </mat-card-subtitle>
+        </mat-card-header>
+        <mat-card-content>
+          <button
+            *ngIf="enableSponsorButton()"
```
Comment (xflord) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html:86]: Do not call functions from the template, as it is inefficient. Either use a pipe or create a property (something like `sponsorButtonEnabled`) that gets updated whenever sponsors are changed.

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html [hunk 6]
```diff
@@ -68,17 +68,29 @@ <h1 class="page-subtitle">{{'MEMBER_DETAIL.OVERVIEW.RESET_PASSWORD' | translate}
...
+            (click)="sponsorThisMember()"
+            class="action-button"
+            color="accent"
+            mat-flat-button>
+            {{'MEMBER_DETAIL.OVERVIEW.SPONSOR_MEMBER'|translate}}
+          </button>
+          <div *ngIf="member.sponsored" class="mt-3">
             <h6 class="d-flex">
               {{'MEMBER_DETAIL.OVERVIEW.SPONSORS' | translate}}:
```
Comment (xflord) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html:95]: Please consider aligning the header with the edit icon as part of this task, it looks bad
Screenshot from 2023-07-10 12-36-05
Also 'This member is sponsored' followed by a 'Sponsor this member' button looks a bit weird. I feel like this sentence is unnecessary now. What do you think @bodnara  @sarkapalkovicova ?

File: apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.components.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
+<div class="{{theme}}">
+  <h1 mat-dialog-title>{{'DIALOGS.SPONSOR_THIS_MEMBER.TITLE' | translate}}</h1>
+  <div mat-dialog-content>
+    <app-choose-sponsor
```
Comment (xflord) [inline apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.components.html:4]: Please make this component's 'Choose sponsor for created user' header more generic. We are not creating a new user in this context (it makes sense only in the dialog for creating a new sponsored member, for that matter)
Screenshot from 2023-07-10 12-39-53

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts [hunk 8]
```diff
@@ -144,6 +160,44 @@ export class MemberOverviewComponent implements OnInit {
...
+        this.loading = true;
+        this.membersService.getRichMemberWithAttributes(this.member.id).subscribe((member) => {
+          this.member = member;
+          this.usersManager.getSponsorsForMember(this.member.id, null).subscribe((sponsors) => {
+            this.sponsors = sponsors;
+            this.sponsorsDataSource.data = this.sponsors;
+          });
+          this.loading = false;
+        });
+      }
```
Comment (xflord) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts:187]: This chunk of code is duplicate, please extract this logic and move it to a separate method (`refreshData()` seems like a good place)

File: apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.component.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    this.theme = this.data.theme;
+    this.member = this.data.member;
+    this.sponsors = this.data.sponsors;
+  }
+
+  onCancel(): void {
+    this.dialogRef.close(false);
+  }
+
+  onSubmit(): void {
```
Comment (xflord) [inline apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.component.ts:50]: I believe there was an effort to stop using the 'onAction()' notation. Not sure though, please ask @bodnara  or @HejdaJakub .

**Review by xflord — CHANGES_REQUESTED** [review]

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html (hunk 6 above)
Comment (Kuliak) [reply apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html:95]: Agree, showing current sponsors should be sufficient, or?

File: apps/admin-gui/src/app/shared/components/choose-sponsor/choose-sponsor.component.ts [hunk 10]
```diff
@@ -13,6 +13,7 @@ export class ChooseSponsorComponent implements OnInit, OnChanges {
   @Input() voSponsors: RichUser[] = [];
   @Input() copy = false;
   @Input() disableSelf = false;
+  @Input() customTitle = '';
```
Comment (xflord) [inline apps/admin-gui/src/app/shared/components/choose-sponsor/choose-sponsor.component.ts:16]: A default title would be nice, just to be safe

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts [hunk 11]
```diff
@@ -138,12 +158,43 @@ export class MemberOverviewComponent implements OnInit {
...
+      sponsors: this.voSponsors.filter(
+        (s) => !this.sponsors.map((sponsor) => sponsor.user.id).includes(s.id)
+      ),
+    };
+
+    const dialogRef = this.dialog.open(SponsorThisMemberDialogComponent, config);
+
+    dialogRef.afterClosed().subscribe((edited) => {
+      if (edited) {
+        this.refreshData();
```
Comment (xflord) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts:183]: Please do this change in the `changeSponsors()` method as well.

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts [hunk 12]
```diff
@@ -100,8 +117,11 @@ export class MemberOverviewComponent implements OnInit {
               this.refreshData();
             }
           },
-          () => (this.loading = false)
-        );
+          error: () => {
+            this.loading = false;
+          },
+        });
+        this.sponsorButtonEnabled = this.isSponsorButtonEnabled();
```
Comment (xflord) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts:124]: Is this necessary? We already call `refreshData()`

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts [hunk 13]
```diff
@@ -35,10 +42,14 @@ export class MemberOverviewComponent implements OnInit {
...
   displayedColumns = ['attName', 'attValue'];
   sponsors: Sponsor[] = [];
+  voSponsors: RichUser[] = [];
   sponsorsDataSource = new MatTableDataSource<Sponsor>();
   vo: Vo;
   loading = false;
   pwdResetAuth: boolean;
+  isSponsor = false;
+  isPerunAdmin = false;
+  sponsorButtonEnabled = true;
```
Comment (xflord) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.ts:52]: Nitpick but false as a default value makes more sense to me

File: apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.component.ts (hunk 9 above)
Comment (bodnara) [reply apps/admin-gui/src/app/shared/components/dialogs/sponsor-this-member-dialog/sponsor-this-member-dialog.component.ts]: I believe in the Angular best practices it is advised  that **Output** of components should not be prefixed with 'on' so it does not collide with vanilla js e.g. `onClick`.
 `<e (something)="onSomething()">`  or `<e (something)="something()">`  should be ok
 `<e (onSomething)="whatever()">` will probably not break anything but preferably don't use this notation

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html (hunk 6 above)
Comment (bodnara) [reply apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html:92]: The design for this use case does include these texts. However it does not seem to add that much value.
image
image

File: apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html [hunk 14]
```diff
@@ -105,9 +106,17 @@ <h6 class="d-flex">
               <tr *matHeaderRowDef="['sponsors', 'expiration']" mat-header-row></tr>
               <tr *matRowDef="let sponsor; columns: ['sponsors', 'expiration'];" mat-row></tr>
             </table>
-          </mat-card-content>
-        </mat-card>
-      </div>
+          </div>
+          <button
```
Comment (xflord) [inline apps/admin-gui/src/app/vos/pages/member-detail-page/member-overview/member-overview.component.html:110]: Could you please add a tooltip when the button is disabled (similar to how it is on the sponsored members page). Information whether the user is not a sponsor/no sponsors are available should be easily retrievable in the `isSponsorButtonEnabled` method.

**Review by xflord — COMMENTED** [review]
Looks pretty good, one last thing

File: apps/admin-gui/src/app/shared/components/choose-sponsor/choose-sponsor.component.ts [hunk 15]
```diff
@@ -13,6 +13,7 @@ export class ChooseSponsorComponent implements OnInit, OnChanges {
   @Input() voSponsors: RichUser[] = [];
   @Input() copy = false;
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/choose-sponsor/choose-sponsor.component.ts:14]: Seems to be unused.
```suggestion
```

**Review by bodnara — APPROVED** [review]
Looks good.

**Pull request merged**